pylint/checkers/variables.py

```diff
@@ -979,267 +992,261 @@ def visit_assignname(self, node: nodes.AssignName) -> None:
             "undefined-variable"
         )
-        used_before_assignment_is_enabled = self.linter.is_message_enabled(
+        self._used_before_assignment_is_enabled = self.linter.is_message_enabled(
             "used-before-assignment"
         )
 
         # iterates through parent scopes, from the inner to the outer
         base_scope_type = self._to_consume[start_index].scope_type
-        # pylint: disable=too-many-nested-blocks; refactoring this block is a pain.
+
         for i in range(start_index, -1, -1):
             current_consumer = self._to_consume[i]
 
-            # The list of base classes in the class definition is not part
-            # of the class body.
-            # If the current scope is a class scope but it's not the inner
-            # scope, ignore it. This prevents to access this scope instead of
-            # the globals one in function members when there are some common
-            # names.
-            if current_consumer.scope_type == "class" and (
-                utils.is_ancestor_name(current_consumer.node, node)
-                or (i != start_index and self._ignore_class_scope(node))
-            ):
-                continue
-
-            # Ignore inner class scope for keywords in class definition
-            if (
-                current_consumer.scope_type == "class"
-                and isinstance(node.parent, nodes.Keyword)
-                and isinstance(node.parent.parent, nodes.ClassDef)
-            ):
+            # Certain nodes shouldn't be checked as they get checked another time
+            if self._node_should_be_skipped(current_consumer, node, i == start_index):
                 continue
 
-            # if the name node is used as a function default argument's value or as
-            # a decorator, then start from the parent frame of the function instead
-            # of the function frame - and thus open an inner class scope
-            if (
-                current_consumer.scope_type == "function"
-                and self._defined_in_function_definition(node, current_consumer.node)
-            ):
-                # ignore function scope if is an annotation/default/decorator, as not in the body
-                continue
+            action, found_nodes = self._check_consumer(
+                node, stmt, frame, current_consumer, i, base_scope_type
+            )
 
-            if current_consumer.scope_type == "lambda" and utils.is_default_argument(
-                node, current_consumer.node
-            ):
+            if action is VariableVisitConsumerAction.CONTINUE:
                 continue
+            if action is VariableVisitConsumerAction.CONSUME:
+                current_consumer.mark_as_consumed(node.name, found_nodes)
+            if action in {
+                VariableVisitConsumerAction.RETURN,
+                VariableVisitConsumerAction.CONSUME,
+            }:
+                return
```

Comment:
I just rebased (locally) #5402 on top of this, and this is the only significant change in the implementation, I just have to plug in here to consume all the uncertain nodes:

```python
            if action in {
                VariableVisitConsumerAction.RETURN,
                VariableVisitConsumerAction.CONSUME,
            }:
                try:
                    uncertain_nodes = current_consumer.consumed_uncertain[node.name]
                except KeyError:
                    uncertain_nodes = []
                current_consumer.mark_as_consumed(node.name, uncertain_nodes)
                return
```

Reply:
Ah, there were failing test cases. And this didn't look right. This looks better for my rebase and passes the tests but requires a change in the PR:

```python
            if action is VariableVisitConsumerAction.CONSUME:
                try:
                    uncertain_nodes = current_consumer.consumed_uncertain[node.name]
                except KeyError:
                    uncertain_nodes = []
                current_consumer.mark_as_consumed(node.name, found_nodes + uncertain_nodes)
```

Reply:
Ah, no, the change can wait for the #5402. I can just make the change in the previous comment _and_ the following change in #5402:

```python
        if not found_nodes:
            self.add_message("used-before-assignment", args=node.name, node=node)
            if node.name in current_consumer.consumed_uncertain:
                uncertain_nodes = current_consumer.consumed_uncertain[node.name]
                return (VariableVisitConsumerAction.CONSUME, found_nodes + uncertain_nodes)
            return (VariableVisitConsumerAction.RETURN, found_nodes)
```